### ISSUE TITLE
Camera Permission Problem on S7 / Android 23 / Android M

### ISSUE THREAD

**Issue by OptrixAU** [artoolkit/arunity5#84]
This is a potential cause of issue #78 

The Android permission model is changing - rather than requesting permissions for every feature your application requires at install-time, the system instead requests permission from the user the _first time you use the feature_

Perhaps the camera isn't working on the Galaxy S7 and above because of an incompatibility along this line - the package may not be correctly asking for permission to use the camera?

**Comment by OptrixAU** [comment]
CONFIRMED AS A PERMISSION ISSUE.

These are the lines immediately before the app fails to start the camera, on 

ServiceManager: Permission failure: android.permission.CAMERA from uid=10235 pid=10605
CameraService: Permission Denial: can't use the camera pid=10605, uid=10235
CameraBase: An error occurred while connecting to camera: 0 

An exception is being raised where **Camera.open** is being called due to the new permission model.

I'm experimenting with requesting the camera permission at runtime during startup of the AR activity. 

To do this properly unfortunately needs us to add the Android support-v4 .arr file from 23.0.0+, but it's currently compiling - the big test on the S7 will happen tomorrow morning.

**Comment by OptrixAU** [comment]
This problem is confirmed and a fix for this issue is included in pull request from https://github.com/OptrixAU/arunity5

**Comment by damonbla** [comment]
Hey there, I'm still having problems getting the camera to work on my Galaxy S7. I can run it and it works great right on my laptop with my webcam, but building for the Android does not work. The problem seems to be related to this issue with S7 phones.

I've done what you mentioned on the artoolkit forums, which was to copy the Android directory from Plugins for the fixed repo to the code I downloaded from artoolkit.org. This did not fix it for me, so I was wondering what I might be able to do beyond just that.

In Android Studio when I use the device monitor, I can see quite a lot of things flying by in the error logs, but mostly it's "isCameraEnabled" which has the "EDMNativeHelperService" tag going by thousands of times. And every once in a while, this error:

10-17 16:35:24.320: E/(1044): Camera[DBG] af_port_ext_send_legacy_status_update: 375: af_port_ext_send_legacy_status_update: Send AF Status update to af port: 4

Any ideas? Thank you!

**Comment by OptrixAU** [comment]
I'm pretty sure that this _specific_ problem (compatibility with the Anrdroid 6.0 permission model) has been resolved. It sounds like you've got a different issue.

The changes I made have already been integrated into the project, so don't bother with my version anymore. Download the proper ARUnity5 and copy the Android files in from there (they don't always rebuild the actual .unitypackage, so it's usually best to manually copy the plugin files across.)

This will also have some fixes for other bugs (like the OpenGL texture bug - if you had the OpenGL texture copying option turned on in my build, you'd only get distorted garbage on your screen) - it may fix your issue. 

Otherwise, open a **new** issue and throw in a little more of your log file - we will try and take a look.

**Comment by damonbla** [comment]
It looks like the problem I'm having still occurs, even with the newest Android files. All I end up with is the default Unity scene (the gray ground and blue sky with the horizon) and nothing else. I'll get some stuff out of my log file and post it to a new issue.

Also I wanted to say thank you for this amazing software! You've made it so easy to start up an AR project. It's fantastic.

**Comment by damonbla** [comment]
Actually I just figured it out! The problem was with the DLLs. I was building for Android, so I just removed all the directories in the Plugins folder that were not Android. This caused DLL loading problems that I didn't notice originally, especially since it still built just fine and actually ran on my phone.

Turns out I had to copy back the folder for my computer (in this case it was x86_64) and just remove the libARWrapper.so file from that directory to get rid of the conflict with the one in the Android directory. And all worked!

Thanks again for the great software and for getting back to me so quickly. Now I get to do some fun AR experimentation.
